src/v/storage/batch_cache.cc

```diff
@@ -9,14 +9,59 @@
 
 #include "batch_cache.h"
 
+#include "model/adl_serde.h"
 #include "utils/to_string.h"
 #include "vassert.h"
 
 #include <fmt/ostream.h>
 
 namespace storage {
 
-batch_cache::entry_ptr
+model::record_batch batch_cache::range::batch(size_t o) {
+    vassert(_valid, "cannot access invalided batch");
+    auto hdr = reflection::from_iobuf<model::record_batch_header>(
+      _arena.share(o, serialized_header_size));
+    auto buffer = _arena.share(
+      o + serialized_header_size,
+      hdr.size_bytes - model::packed_record_batch_header_size);
+
+    return model::record_batch(
+      hdr, std::move(buffer), model::record_batch::tag_ctor_ng{});
+}
+
+size_t batch_cache::range::memory_size() const {
+    if (_arena.empty()) {
+        return 0;
+    }
+    return _arena.begin()->capacity();
+}
+size_t batch_cache::range::bytes_left() const {
+    return _arena.begin()->available_bytes();
+}
+bool batch_cache::range::empty() const {
+    return bytes_left() == _arena.size_bytes();
+}
+bool batch_cache::range::full() const { return bytes_left() == 0; }
+
+bool batch_cache::range::fits(const model::record_batch& b) const {
+    return bytes_left() >= (serialized_header_size + b.data().size_bytes());
```

Comment:
for a workload where batches contain `64K` payloads (I dunno, like an image) then the batch size will be something like `64K+X` where X is the header and a little overhead. Would this `fits` rule then waste 64K bytes per batch? I think that it may be worth while to think about packing data into left over space and extending the iobuf with one or more smaller fragments (will happen automatically in range::add during append). Maybe we'd add a threshold or something, but it does seem like we may encounter an adversarial workload with larger batches. I recall I asked this on the review this morning and the concern was then smaller allocations. But that seems like a good trade off, right? We reduce those by some factor dependent on how many we can pack on average into an arena?

Reply:
This made me thinking and testing it more. I think the suggestion is valid. I am going to send implementation where waste percentage in range is bounded.